src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java

```diff
@@ -0,0 +1,128 @@
+import software.amazon.awssdk.eventstreamrpc.OperationContinuationHandlerContext;
+import software.amazon.awssdk.eventstreamrpc.model.EventStreamJsonMessage;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.aws.greengrass.ipc.common.ExceptionUtil.translateExceptions;
+import static software.amazon.awssdk.aws.greengrass.GreengrassCoreIPCService.GET_CLIENT_DEVICE_AUTH_TOKEN;
+
+public class GetClientDeviceAuthTokenOperationHandler
+        extends GeneratedAbstractGetClientDeviceAuthTokenOperationHandler {
+    private static final Logger logger = LogManager.getLogger(GetClientDeviceAuthTokenOperationHandler.class);
+    private static final String COMPONENT_NAME = "componentName";
+    private static final String UNAUTHORIZED_ERROR = "Not Authorized";
+    private static final String MQTT_CREDENTIAL_TYPE = "mqtt";
+    private static final String NO_DEVICE_CREDENTIAL_ERROR = "Client device credential document is required";
+    private static final String NO_VALID_MQTT_CREDENTIAL_ERROR = "Valid mqtt credential document is required";
+    private final String serviceName;
+    private final AuthorizationHandler authorizationHandler;
+    private final SessionManager sessionManager;
+    private final Map<String, String> credentialMap = new HashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param context              operation continuation handler
+     * @param sessionManager       session manager
+     * @param authorizationHandler authorization handler
+     */
+    public GetClientDeviceAuthTokenOperationHandler(
+            OperationContinuationHandlerContext context,
+            SessionManager sessionManager,
+            AuthorizationHandler authorizationHandler
+    ) {
+
+        super(context);
+        serviceName = context.getAuthenticationData().getIdentityLabel();
+        this.sessionManager = sessionManager;
+        this.authorizationHandler = authorizationHandler;
+    }
+
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.PreserveStackTrace"})
+    @Override
+    public GetClientDeviceAuthTokenResponse handleRequest(GetClientDeviceAuthTokenRequest request) {
+        return translateExceptions(() -> {
+            try {
+                doAuthorizationForClientDevAuthToken();
+            } catch (AuthorizationException e) {
+                logger.atWarn().kv("error", e.getMessage()).kv(COMPONENT_NAME, serviceName).log(UNAUTHORIZED_ERROR);
+                throw new UnauthorizedError(e.getMessage());
+            }
+            Map<String, String> credentialMap = mapOfMqttCredential(request.getCredential());
+            try {
+                String sessionId = sessionManager.createSession(MQTT_CREDENTIAL_TYPE, credentialMap);
+                GetClientDeviceAuthTokenResponse response = new GetClientDeviceAuthTokenResponse();
+                return response.withClientDeviceAuthToken(sessionId);
+            } catch (Exception e) {
```

Comment:
if it throws a more specific error and we can give the user a better indication of what went wrong, then we should do that. I just want to be complete and make sure you additionally have a catch all.

Reply:
AuthenticationException should be treated differently - we return false to the client, rather than throwing the client an exception right

Reply:
ok

Reply:
But it isn't a ServiceError, it should be a properly modeled exception in that case.

Reply:
`InvalidCredentialError` is what is in the design doc

Reply:
+1 The session manager throws an `AuthenticationException` which should be caught and translated here